### PR TITLE
Testing renter API functions

### PR DESCRIPTION
**Pull request by mnsl**
This PR expands `TestIntegrationHostAndRent` in order to attain complete coverage of renter API calls (excluding calls that rely on sharing, a feature that is currently disabled).

`TestIntegrationHostAndRent` is now unwieldy, but breaking the tests into separate functions wastes several seconds per function because of setup.

Renter testing is still incomplete, of course. We need to be doing tests that include multiple hosts and renters in realistic (out of sync) conditions.

### CODE REVIEW

File: api/renterhost_test.go [hunk 1]
```diff
@@ -44,30 +47,130 @@ func TestIntegrationHostAndRent(t *testing.T) {
...
+	if got := get.FinancialMetrics.ContractSpending; got.Cmp(expectedContractSpending) != 0 {
+		t.Fatalf("expected contract spending to be %v; got %v", expectedContractSpending, got)
+	}
+
+	// Check that renterHandlerGET correctly handles invalid inputs.
+	// Try an empty funds string.
+	allowanceValues.Set("funds", "")
+	allowanceValues.Set("period", testPeriod)
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err.Error() != "Couldn't parse funds" {
```
Comment (lukechampine) [inline api/renterhost_test.go]: these are unsafe, because `err` can be nil.

File: api/renterhost_test.go [hunk 2]
```diff
@@ -44,30 +47,130 @@ func TestIntegrationHostAndRent(t *testing.T) {
...
+	allowanceValues.Set("funds", "0")
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err.Error() != contractor.ErrInsufficientAllowance.Error() {
+		t.Errorf("expected error to be %v; got %v", contractor.ErrInsufficientAllowance, err)
+	}
+	// Try a empty period string.
+	allowanceValues.Set("funds", testFunds)
+	allowanceValues.Set("period", "")
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err.Error()[:23] != "Couldn't parse period: " {
```
Comment (lukechampine) [inline api/renterhost_test.go]: less brittle would be `strings.HasPrefix`.

File: api/renterhost_test.go [hunk 3]
```diff
@@ -90,14 +193,22 @@ func TestIntegrationHostAndRent(t *testing.T) {
...
 	}
 	if len(rf.Files) != 2 || rf.Files[0].UploadProgress < 10 || rf.Files[1].UploadProgress < 10 {
 		t.Fatal("the uploading is not succeeding for some reason:", rf.Files[0], rf.Files[1])
 	}
+	// Try uploading a nonexistent file.
+	path = filepath.Join(st.dir, "fake.dat")
+	uploadValues = url.Values{}
+	uploadValues.Set("source", path)
+	err = st.stdPostAPI("/renter/upload/fake", uploadValues)
+	if n := len(err.Error()); err.Error()[n-25:] != "no such file or directory" {
```
Comment (lukechampine) [inline api/renterhost_test.go]: `strings.HasSuffix`

**Comment by DavidVorick** [comment]
Adding several seconds to testing time is not terrible if it makes the tests more approachable/readable. Especially if we start running nightly builds for the really long tests instead of running all tests every PR.

File: api/renterhost_test.go [hunk 4]
```diff
@@ -44,30 +47,130 @@ func TestIntegrationHostAndRent(t *testing.T) {
...
 	}
 
-	// create a file
+	// The renter should now have 1 contract.
+	if err = st.getAPI("/renter/contracts", &contracts); err != nil {
+		t.Fatal(err)
+	}
+	if len(contracts.Contracts) != 1 {
+		t.Fatalf("expected renter to have 1 contract; got %v", len(contracts.Contracts))
+	}
```
Comment (DavidVorick) [inline api/renterhost_test.go]: Probably worth checking some of the fields inside the contract as well.

File: api/renterhost_test.go [hunk 5]
```diff
@@ -44,30 +47,130 @@ func TestIntegrationHostAndRent(t *testing.T) {
...
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err.Error()[:23] != "Couldn't parse period: " {
+		t.Errorf("expected error to begin with 'Couldn't parse period: '; got %v", err)
+	}
+	// Try a period that will lead to a length-zero RenewWindow.
+	allowanceValues.Set("period", "1")
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err.Error() != contractor.ErrAllowanceZeroWindow.Error() {
+		t.Errorf("expected error to be %v, got %v", contractor.ErrAllowanceZeroWindow, err)
+	}
```
Comment (DavidVorick) [inline api/renterhost_test.go]: Probably worth moving these edge case checks to a different test.

File: api/renterhost_test.go [hunk 6]
```diff
@@ -118,6 +229,38 @@ func TestIntegrationHostAndRent(t *testing.T) {
 		t.Fatal("data mismatch when downloading a file")
 	}
 
+	// Try downloading a nonexistent file.
+	err = st.stdGetAPI("/renter/download/fake?destination=" + downpath)
+	if err.Error() != "Download failed: no file with that path" {
+		t.Errorf("expected error to be 'Download failed: no file with that path'; got %v instead", err)
+	}
```
Comment (DavidVorick) [inline api/renterhost_test.go]: I slammed all of the edge case testing into a single test in the host, and I now regret it because I find it very difficult to look at the test and know what's going on. And if I want to change it, I have to worry about what edge cases I might be changing. (if there's an empty folder test further down I have to move it, for example, but it's also possible to silently make the test useless by setting up conditions where the test will now accidentally always pass).

**Comment by DavidVorick** [comment]
What's the status on this?

**Comment by mnsl** [comment]
On its way, sorry

**Comment by mnsl** [comment]
Latest changes:
* clean up error checking
* check that a file's renter nickname is a nonempty string
* reorganize new testing into smaller tests rather than bloating up TestIntegration

File: api/renter_test.go [hunk 7]
```diff
@@ -244,3 +253,360 @@ func TestRenterHostsAllHandler(t *testing.T) {
 		t.Fatalf("expected 1 host, got %v", len(ah.Hosts))
 	}
 }
+
+func TestRenterHandlerContracts(t *testing.T) {
```
Comment (DavidVorick) [inline api/renter_test.go]: Should have a comment for the function

File: api/renter_test.go [hunk 8]
```diff
@@ -244,3 +253,360 @@ func TestRenterHostsAllHandler(t *testing.T) {
...
+	for _, contract := range contracts.Contracts {
+		expectedContractSpending = expectedContractSpending.Add(contract.RenterFunds)
+	}
+	if got := get.FinancialMetrics.ContractSpending; got.Cmp(expectedContractSpending) != 0 {
+		t.Fatalf("expected contract spending to be %v; got %v", expectedContractSpending, got)
+	}
+
+}
+
+func TestRenterHandlerGetAndPost(t *testing.T) {
```
Comment (DavidVorick) [inline api/renter_test.go]: ditto

File: api/renter_test.go [hunk 9]
```diff
@@ -244,3 +253,360 @@ func TestRenterHostsAllHandler(t *testing.T) {
...
+	}
+	// Try a period that will lead to a length-zero RenewWindow.
+	allowanceValues.Set("period", "1")
+	err = st.stdPostAPI("/renter", allowanceValues)
+	if err == nil || err.Error() != contractor.ErrAllowanceZeroWindow.Error() {
+		t.Errorf("expected error to be %v, got %v", contractor.ErrAllowanceZeroWindow, err)
+	}
+}
+
+func TestRenterLoadNonexistent(t *testing.T) {
```
Comment (DavidVorick) [inline api/renter_test.go]: ditto

File: api/renter_test.go [hunk 10]
```diff
@@ -244,3 +253,360 @@ func TestRenterHostsAllHandler(t *testing.T) {
...
+	var queue RenterDownloadQueue
+	if err = st.getAPI("/renter/downloads", &queue); err != nil {
+		t.Fatal(err)
+	}
+	if len(queue.Downloads) != 0 {
+		t.Fatalf("expected renter to have 0 downloads in the queue; got %v", len(queue.Downloads))
+	}
+}
+
+func TestRenterHandlerRename(t *testing.T) {
```
Comment (DavidVorick) [inline api/renter_test.go]: ditto

File: api/renter_test.go [hunk 11]
```diff
@@ -244,3 +253,360 @@ func TestRenterHostsAllHandler(t *testing.T) {
...
+
+	// Try renaming a nonexistent file.
+	renameValues.Set("newsiapath", "newdne")
+	err = st.stdPostAPI("/renter/rename/dne", renameValues)
+	if err == nil || err.Error() != renter.ErrUnknownPath.Error() {
+		t.Errorf("expected error to be %v; got %v", renter.ErrUnknownPath, err)
+	}
+}
+
+func TestRenterHandlerDelete(t *testing.T) {
```
Comment (DavidVorick) [inline api/renter_test.go]: ditto

**Comment by DavidVorick** [comment]
Just needs some docstrings. Thanks for breaking down the functions more.

**Comment by mnsl** [comment]
Getting 'too many open files' errors on my Mac, which we thought we fixed with the global http.DefaultClient in 1373. The failures occur even when the api package is run on its own.

```
=== RUN   TestRenterHostsAllHandler
--- FAIL: TestRenterHostsAllHandler (0.25s)
	renter_test.go:233: listen tcp: lookup localhost: no such host
=== RUN   TestRenterHandlerContracts
--- FAIL: TestRenterHandlerContracts (0.00s)
	renter_test.go:266: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestRenterHandlerContracts/gateway/gateway.log: too many open files
=== RUN   TestRenterHandlerGetAndPost
--- FAIL: TestRenterHandlerGetAndPost (0.00s)
	renter_test.go:331: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestRenterHandlerGetAndPost/gateway/gateway.log: too many open files
=== RUN   TestRenterLoadNonexistent
--- FAIL: TestRenterLoadNonexistent (0.00s)
	renter_test.go:428: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestRenterLoadNonexistent/gateway/gateway.log: too many open files
=== RUN   TestRenterHandlerRename
--- FAIL: TestRenterHandlerRename (0.00s)
	renter_test.go:486: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestRenterHandlerRename/gateway/gateway.log: too many open files
=== RUN   TestRenterHandlerDelete
--- FAIL: TestRenterHandlerDelete (0.00s)
	renter_test.go:569: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestRenterHandlerDelete/gateway/gateway.log: too many open files
=== RUN   TestIntegrationHostAndRent
--- FAIL: TestIntegrationHostAndRent (0.00s)
	renterhost_test.go:29: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestIntegrationHostAndRent/gateway/gateway.log: too many open files
=== RUN   TestExplorerPreset
--- FAIL: TestExplorerPreset (0.00s)
	server_test.go:16: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestExplorerPreset/gateway/gateway.log: too many open files
=== RUN   TestReloading
--- FAIL: TestReloading (0.00s)
	server_test.go:40: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestReloading/gateway/gateway.log: too many open files
=== RUN   TestAuthentication
--- FAIL: TestAuthentication (0.00s)
	server_test.go:71: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/authenticated-api/TestAuthentication/gateway/gateway.log: too many open files
=== RUN   TestIntegrationWalletGETEncrypted
--- FAIL: TestIntegrationWalletGETEncrypted (0.00s)
	wallet_test.go:29: Failed to create gateway: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestIntegrationWalletGETEncrypted/gateway/gateway.log: too many open files
=== RUN   TestIntegrationWalletBlankEncrypt
--- FAIL: TestIntegrationWalletBlankEncrypt (0.00s)
	wallet_test.go:94: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestIntegrationWalletBlankEncrypt/gateway/gateway.log: too many open files
=== RUN   TestIntegrationWalletGETSiacoins
--- FAIL: TestIntegrationWalletGETSiacoins (0.00s)
	wallet_test.go:156: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestIntegrationWalletGETSiacoins/gateway/gateway.log: too many open files
=== RUN   TestIntegrationWalletTransactionGETid
--- FAIL: TestIntegrationWalletTransactionGETid (0.00s)
	wallet_test.go:250: open /var/folders/ps/z2hlbr517dn0ps9b9wwpbnf00000gn/T/SiaTesting/api/TestIntegrationWalletTransactionGETid/gateway/gateway.log: too many open files
FAIL
FAIL	github.com/NebulousLabs/Sia/api	13.688s
```

**Comment by DavidVorick** [comment]
rebase to master, we pulled the mac build. I'm not sure what the file limit is, but I think it's very low. The ecosystem test I was running was triggering the error regardless of `defer Close` calls. There might be a file leak somewhere, but it seems more likely that the max files on the mac build is under 1000, maybe under 100. 6 nodes simultaneously seems to overwhelm it.

**Comment by mnsl** [comment]
I had rebased; Travis passed, just not my Mac.  Increased the open files limit on my Mac from 256 (the default) to 1024 (`ulimit -n 1024`) and tests started passing again.  The limit currently needs to be more than 367.

**Comment by mnsl** [comment]
Should be ready to merge.

File: modules/renter/contractor/allowance.go [hunk 12]
```diff
@@ -12,8 +12,12 @@ import (
 var (
 	errAllowanceNoHosts    = errors.New("hosts must be non-zero")
 	errAllowanceZeroPeriod = errors.New("period must be non-zero")
-	errAllowanceZeroWindow = errors.New("renew window must be non-zero")
 	errAllowanceWindowSize = errors.New("renew window must be less than period")
+
+	// ErrAllowanceZeroWindow is returned when the caller requests a
+	// zero-length renewal window. This will happen if the caller sets the
+	// period to 1 block, since RenewWindow := period / 2.
```
Comment (lukechampine) [inline modules/renter/contractor/allowance.go]: that detail is currently specific to the API, not the renter module
